rust/xaynet-server/src/metrics/recorders/influxdb/models.rs

```diff
@@ -0,0 +1,259 @@
+use chrono::{DateTime, Utc};
+use influxdb::InfluxDbWriteable;
+
+/// An enum that contains all supported measurements.
+pub enum Measurement {
+    RoundParamSum,
+    RoundParamUpdate,
+    Phase,
+    MasksTotalNumber,
+    RoundTotalNumber,
+    MessageSum,
+    MessageUpdate,
+    MessageSum2,
+    MessageDiscarded,
+    MessageRejected,
+}
+
+impl From<&Measurement> for &'static str {
+    fn from(measurement: &Measurement) -> &'static str {
+        match measurement {
+            Measurement::RoundParamSum => "round_param_sum",
+            Measurement::RoundParamUpdate => "round_param_update",
+            Measurement::Phase => "phase",
+            Measurement::MasksTotalNumber => "masks_total_number",
+            Measurement::RoundTotalNumber => "round_total_number",
+            Measurement::MessageSum => "message_sum",
+            Measurement::MessageUpdate => "message_update",
+            Measurement::MessageSum2 => "message_sum2",
+            Measurement::MessageDiscarded => "message_discarded",
+            Measurement::MessageRejected => "message_rejected",
+        }
+    }
+}
+
+impl ToString for Measurement {
+    fn to_string(&self) -> String {
+        Into::<&str>::into(self).into()
+    }
+}
+
+/// A container that contains the tags of a metric.
+pub struct Tags(Vec<(String, influxdb::Type)>);
+
+impl Tags {
+    /// Creates a new empty container for tags.
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Adds a tag to the metric.
+    pub fn add<K, V>(&mut self, key: K, value: V)
+    where
+        K: Into<String>,
+        V: Into<influxdb::Type>,
+    {
+        self.0.push((key.into(), value.into()))
+    }
+
+    pub(in crate::metrics) fn into_inner(self) -> Vec<(String, influxdb::Type)> {
+        self.0
+    }
+}
+
+impl Default for Tags {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A metrics data point.
+pub(in crate::metrics) struct Metric {
+    name: Measurement,
+    time: DateTime<Utc>,
+    value: influxdb::Type,
+    tags: Option<Tags>,
+}
+
+impl Metric {
+    pub(in crate::metrics) fn new(measurement: Measurement, value: influxdb::Type) -> Self {
+        Self {
+            name: measurement,
+            time: Utc::now(),
+            value,
+            tags: None,
+        }
+    }
+
+    pub(in crate::metrics) fn with_tags(&mut self, tags: Tags) {
+        // we don't want to extend tags. The user can't call that method
```

Comment:
What do you mean by "extending" the tags?

Reply:
With `metric!` It is not possible to call the function several times. Therefore, I did't cover the case where we would extend `self.tags` when `self.tags` already contains tags.
I updated the comment